File: frontend/src/services/api/deliveryEarningService.ts
import { axiosInstance } from "@/services/api/axiosInstance";
import type {
  DeliveryEarningsDashboard,
  DeliveryEarning,
} from "@/types/deliveryEarning.types";

/**
 * Delivery Earning API client.
 * Backend reference: DeliveryEarningController at /api/delivery-earnings.
 * Ownership enforced server-side — delivery boys see only their own.
 */
export const deliveryEarningService = {
  /** Delivery boy: get full earnings dashboard. */
  async getMyDashboard(): Promise<DeliveryEarningsDashboard> {
    const response = await axiosInstance.get<DeliveryEarningsDashboard>(
      "/delivery-earnings/my/dashboard",
    );
    return response.data;
  },

  /** Delivery boy: get order-wise earnings history. */
  async getMyHistory(): Promise<DeliveryEarning[]> {
    const response = await axiosInstance.get<DeliveryEarning[]>(
      "/delivery-earnings/my/history",
    );
    return response.data;
  },

  /** Admin: get aggregated earnings across all delivery boys. */
  async getAdminDashboard(): Promise<DeliveryEarningsDashboard> {
    const response = await axiosInstance.get<DeliveryEarningsDashboard>(
      "/delivery-earnings/admin/dashboard",
    );
    return response.data;
  },

  /** Admin: get detailed earnings for a specific delivery boy. */
  async getDeliveryBoyDashboard(
    deliveryBoyId: number,
  ): Promise<DeliveryEarningsDashboard> {
    const response = await axiosInstance.get<DeliveryEarningsDashboard>(
      `/delivery-earnings/admin/${deliveryBoyId}/dashboard`,
    );
    return response.data;
  },

  /** Admin: get order-wise earnings for a specific delivery boy. */
  async getDeliveryBoyHistory(
    deliveryBoyId: number,
  ): Promise<DeliveryEarning[]> {
    const response = await axiosInstance.get<DeliveryEarning[]>(
      `/delivery-earnings/admin/${deliveryBoyId}/history`,
    );
    return response.data;
  },
};

File: backend/src/main/java/com/distributrack/service/DeliveryEarningService.java
package com.distributrack.service;

import com.distributrack.dto.response.DeliveryEarningsDashboard;
import com.distributrack.dto.response.DeliveryEarningResponse;
import com.distributrack.entity.Delivery;

import java.util.List;

public interface DeliveryEarningService {

    /**
     * Calculate and save earning for a completed delivery.
     * Called when delivery status transitions to DELIVERED.
     * Safe to call multiple times — duplicate deliveries are rejected.
     */
    DeliveryEarningResponse createEarningIfNotExists(Delivery delivery);

    /**
     * Delivery boy: get full earnings dashboard (today, month, history).
     */
    DeliveryEarningsDashboard getMyEarningsDashboard();

    /**
     * Admin: get aggregated earnings dashboard for all delivery boys.
     */
    DeliveryEarningsDashboard getAdminEarningsDashboard();

    /**
     * Admin: get detailed earnings dashboard for a specific delivery boy.
     */
    DeliveryEarningsDashboard getDeliveryBoyEarningsDashboard(Long deliveryBoyId);

    /**
     * Delivery boy: get order-wise earnings history.
     */
    List<DeliveryEarningResponse> getMyEarningsHistory();

    /**
     * Admin: get all earnings for a specific delivery boy.
     */
    List<DeliveryEarningResponse> getDeliveryBoyEarningsHistory(Long deliveryBoyId);
}
